pkg/scheduler/internal/queue/scheduling_queue.go

```diff
@@ -554,24 +569,43 @@ func (p *PriorityQueue) runPreEnqueuePlugin(ctx context.Context, pl framework.Pr
 	return s
 }
 
-// addToActiveQ tries to add pod to active queue. It returns 2 parameters:
+// moveToActiveQ tries to add pod to active queue and remove it from unschedulable and backoff queues.
+// It returns 2 parameters:
 // 1. a boolean flag to indicate whether the pod is added successfully.
 // 2. an error for the caller to act on.
-func (p *PriorityQueue) addToActiveQ(logger klog.Logger, pInfo *framework.QueuedPodInfo) (bool, error) {
+func (p *PriorityQueue) moveToActiveQ(logger klog.Logger, pInfo *framework.QueuedPodInfo, event string) (bool, error) {
+	gatedBefore := pInfo.Gated
 	pInfo.Gated = !p.runPreEnqueuePlugins(context.Background(), pInfo)
+
+	p.activeQLock.Lock()
+	defer p.activeQLock.Unlock()
 	if pInfo.Gated {
 		// Add the Pod to unschedulablePods if it's not passing PreEnqueuePlugins.
+		if _, exists, _ := p.activeQ.Get(pInfo); exists {
+			return false, nil
+		}
+		if _, exists, _ := p.podBackoffQ.Get(pInfo); exists {
+			return false, nil
+		}
 		p.unschedulablePods.addOrUpdate(pInfo)
 		return false, nil
 	}
 	if pInfo.InitialAttemptTimestamp == nil {
 		now := p.clock.Now()
 		pInfo.InitialAttemptTimestamp = &now
 	}
+
 	if err := p.activeQ.Add(pInfo); err != nil {
 		logger.Error(err, "Error adding pod to the active queue", "pod", klog.KObj(pInfo.Pod))
 		return false, err
 	}
+
+	p.unschedulablePods.delete(pInfo.Pod, gatedBefore)
+	_ = p.podBackoffQ.Delete(pInfo) // Don't need to react when pInfo is not found.
+	logger.V(5).Info("Pod moved to an internal scheduling queue", "pod", klog.KObj(pInfo.Pod), "event", event, "queue", activeQ)
+	metrics.SchedulerQueueIncomingPods.WithLabelValues("active", event).Inc()
+	p.addNominatedPodUnlocked(logger, pInfo.PodInfo, nil)
```

Comment:
Do we actually have to run `addNominatedPodUnlocked` every time moving the Pod to activeQ? Looks like unnecessary unless the Pod is just `Update()`-ed or `Add()`-ed.

Reply:
Good point. Changed to call `addNominatedPodUnlocked` only for `PodAdd` and `PodUpdate` events. Before, it was also called for `ForceActivate` event, but I think it's redundant, as the pod is already in one of the queues, so it's in the nominator as well.